In [13]:
#importing data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df = pd.read_csv("StudentsPerformance.csv")
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group C            some high school  free/reduced   
1  female        group B                 high school  free/reduced   
2  female        group B            some high school  free/reduced   
3  female        group B                some college      standard   
4  female        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none           0             17             10  
1                    none           8             24             23  
2                    none          18             32             28  
3                    none          11             38             32  
4                    none          22             39             33

In [15]:
#checking null values
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

# 1. Find out how many males and females participated in the test.

In [16]:
df["gender"].value_counts()

female    518
male      482
Name: gender, dtype: int64

### 518 female stuedents and 482 male students are participated in the test

# 2. What do you think about the students' parental level of education.

In [28]:
df['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

### 1.Majority of the parent having good educational qualification
### 2.59 parents have master's degree
### 3. Few of them with high school level education

# 3.Who scores the most on average for math, reading and writing based on
## (a) Gender
#### i) most on average for math

In [30]:
df1 = df.groupby('gender')['math score'].mean()
df1

gender
female    63.376448
male      68.821577
Name: math score, dtype: float64

#### ii) most on average for reading

In [31]:
df2 = df.groupby('gender')['reading score'].mean()
df2

gender
female    72.590734
male      65.545643
Name: reading score, dtype: float64

#### iii) most on average for writing

In [32]:
df3 = df.groupby('gender')['writing score'].mean()
df3

gender
female    72.467181
male      63.446058
Name: writing score, dtype: float64

### Boys scored higher in mathematics whereas female students scored higher in reading and writing 

## (b) Test preparation course
#### i) most on average for math

In [42]:
df4 = df.groupby('test preparation course')['math score'].mean()
df4

test preparation course
completed    69.96648
none         63.78972
Name: math score, dtype: float64

#### ii) most on average for reading

In [44]:
df5 = df.groupby('test preparation course')['reading score'].mean()
df5

test preparation course
completed    74.175978
none         66.417445
Name: reading score, dtype: float64

#### iii) most on average for writing

In [46]:
df6 = df.groupby('test preparation course')['writing score'].mean()
df6

test preparation course
completed    74.684358
none         64.457944
Name: writing score, dtype: float64

### Students who have completed the test preparation course are excelled in all the exams

# 4.What do you think about the scoring variation for math, reading and writing based on
## (a) Gender


In [50]:
df.groupby(['gender']).agg(['min','median','max','mean'])

math score                        reading score                         \
              min median  max       mean           min median  max       mean   
gender                                                                          
female          0     65  100  63.376448            17     73  100  72.590734   
male           27     69  100  68.821577            23     66  100  65.545643   

       writing score                         
                 min median  max       mean  
gender                                       
female            10     74  100  72.467181  
male              15     64  100  63.446058

### Female students performed well in reading and writing while male students in math

## (b) Test preparation course

In [51]:
df.groupby(['test preparation course']).agg(['min','median','max','mean'])

math score                       reading score         \
                               min median  max      mean           min median   
test preparation course                                                         
completed                       23     70  100  69.96648            37     75   
none                             0     64  100  63.78972            17     67   

                                        writing score                         
                         max       mean           min median  max       mean  
test preparation course                                                       
completed                100  74.175978            36     76  100  74.684358  
none                     100  66.417445            10     65  100  64.457944

### In all subject ,those who completed test preparation course performed well than the others

## 5. The management needs your help to give bonus points to the top 25% of students based on their maths score, so how will you help the management to achieve this

In [54]:
mh = df.sort_values(['math score'], ascending=False)
mh.head(int(len(mh)*(25/100)))

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
856    male        group E            some high school      standard   
855    male        group E          associate's degree  free/reduced   
854    male        group D            some high school      standard   
853    male        group D          associate's degree  free/reduced   
852    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
856               completed          77             76             77  
855               completed          77             69             68  
854               completed          77             68             69  
853                    none          77             78             73  
852                    none          77             62             64  

[250 rows x 8 columns]